clippy_lints/src/needless_ok.rs

```diff
@@ -0,0 +1,129 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::method_chain_args;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::is_type_diagnostic_item;
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, MatchSource, PatKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///* Checks for unnecessary `ok()` in if let.
+    ///
+    /// ### Why is this bad?
+    /// Calling `ok()` in if let is unnecessary, instead match
+    /// on `Ok(pat)`
+    ///
+    /// ### Example
+    /// ```ignore
+    /// for i in iter {
+    ///     if let Some(value) = i.parse().ok() {
+    ///         vec.push(value)
+    ///     }
+    /// }
+    /// ```
+    /// Could be written:
+    ///
+    /// ```ignore
+    /// for i in iter {
+    ///     if let Ok(value) = i.parse() {
+    ///         vec.push(value)
+    ///     }
+    /// }
+    /// ```
+    pub IF_LET_SOME_RESULT,
+    style,
+    "usage of `ok()` in `if let Some(pat)` statements is unnecessary, match on `Ok(pat)` instead"
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for unnecessary `ok()` in `while let`.
+    ///
+    /// ### Why is this bad?
+    /// Calling `ok()` in `while let` is unnecessary, instead match
+    /// on `Ok(pat)`
+    ///
+    /// ### Example
+    /// ```ignore
+    /// while let Some(value) = iter.next().ok() {
+    ///     vec.push(value)
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```ignore
+    /// while let Ok(value) = iter.next() {
+    ///     vec.push(value)
+    /// }
+    /// ```
+    pub WHILE_LET_SOME_RESULT,
+    style,
+    "usage of `ok()` in `while let Some(pat)` statements is unnecessary, match on `Ok(pat)` instead"
+}
+
+declare_lint_pass!(NeedlessOk => [IF_LET_SOME_RESULT, WHILE_LET_SOME_RESULT]);
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessOk {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! { // IF_LET_SOME_RESULT check
+            if let ExprKind::Match(op, body, MatchSource::IfLetDesugar { .. }) = expr.kind; // test if expr is `if let`
```

Comment:
these two if_chains should be merged, they're basically identical except for the very first part, and you can move that out of the if_chain

Reply:
Good call!

Reply:
(any updates?)

Reply:
See other thread.